src/SegmentationCameraSensor.cc

```diff
@@ -0,0 +1,511 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include <mutex>
+#include <memory>
+
+#include <ignition/common/Console.hh>
+#include <ignition/common/Image.hh>
+#include <ignition/common/Profiler.hh>
+#include <ignition/msgs.hh>
+#include <ignition/rendering/SegmentationCamera.hh>
+#include <ignition/transport/Node.hh>
+#include <ignition/transport/Publisher.hh>
+
+#include "ignition/sensors/RenderingEvents.hh"
+#include "ignition/sensors/SegmentationCameraSensor.hh"
+#include "ignition/sensors/SensorFactory.hh"
+
+using namespace ignition;
+using namespace sensors;
+
+class ignition::sensors::SegmentationCameraSensorPrivate
+{
+  /// \brief Save an image
+  /// \param[in] _coloredBuffer buffer of colored map
+  /// \param[in] _labelsBuffer buffer of labels map
+  /// \param[in] _width width of image in pixels
+  /// \param[in] _height height of image in pixels
+  /// \return True if the image was saved successfully. False can mean
+  /// that the path provided to the constructor does exist and creation
+  /// of the path was not possible.
+  /// \sa ImageSaver
+  public: bool SaveImage(const uint8_t *_coloredBuffer,
+    const uint8_t *_labelsBuffer, unsigned int _width, unsigned int _height);
+
+  /// \brief SDF Sensor DOM Object
+  public: sdf::Sensor sdfSensor;
+
+  /// \brief True if Load() has been called and was successful
+  public: bool initialized = false;
+
+  /// \brief Rendering Segmentation Camera
+  public: rendering::SegmentationCameraPtr camera {nullptr};
+
+  /// \brief Node to create publisher
+  public: transport::Node node;
+
+  /// \brief Publisher to publish segmentation colored image
+  public: transport::Node::Publisher coloredMapPublisher;
+
+  /// \brief Publisher to publish segmentation labels image
+  public: transport::Node::Publisher labelsMapPublisher;
+
+  /// \brief Segmentation colored image message
+  public: msgs::Image coloredMapMsg;
+
+  /// \brief Segmentation labels image message
+  public: msgs::Image labelsMapMsg;
```

Comment:
Do these messages need to be data members of the `SegmentationCameraSensorPrivate` class? It looks like the're only used in the `Update` method, so I think it would be better to create new message variables in that method instead.

Reply:
That is to avoid creating them each time in Update as it is called periodically